From 24dc4ff807752e82b8e8d57b04a8b86359674cff Mon Sep 17 00:00:00 2001
From: Jason Hu <jason.hu@intel.com>
Date: Thu, 26 Jul 2012 17:25:57 -0400
Subject: [PATCH] [PORT FROM MAIN] To fix video flicking issue when rendering
 through overlay.

Currently, hw overlay need always hold an output buffer until flip
to another buffer. So we should hold two more buffers in surface
texture back end to avoid blocking on dequeue buffer by decoder.

Change-Id: I07a6b40b7664a37e9549e43146969946ca3006e0
Category: feature differentiation
Domain: Video.Rendering-HDMI
Origin: internal
Signed-off-by: Jason Hu <jason.hu@intel.com>
Signed-off-by: Dan Liang <dan.liang@intel.com>.

diff --git a/include/media/stagefright/OMXCodec.h b/include/media/stagefright/OMXCodec.h
index 9b7a656..1d76c6d 100644
--- a/include/media/stagefright/OMXCodec.h
+++ b/include/media/stagefright/OMXCodec.h
@@ -110,6 +110,7 @@ struct OMXCodec : public MediaSource,
 #ifdef MTK_HARDWARE
         kAvoidMemcopyInputRecordingFrames     = 8192,
 #endif
+        kRequiresHoldExtraBuffers             = 65536,
 #if defined(OMAP_ENHANCEMENT)
         kAvoidMemcopyInputRecordingFrames     = 0x20000000,
 #endif
diff --git a/media/libstagefright/OMXCodec.cpp b/media/libstagefright/OMXCodec.cpp
index 269028a..0f58d48 100644
--- a/media/libstagefright/OMXCodec.cpp
+++ b/media/libstagefright/OMXCodec.cpp
@@ -406,6 +406,10 @@ uint32_t OMXCodec::getComponentQuirks(
                 index, "defers-output-buffer-allocation")) {
         quirks |= kDefersOutputBufferAllocation;
     }
+    if (list->codecHasQuirk(
+                index, "requires-hold-extra-buffers")) {
+        quirks |= kRequiresHoldExtraBuffers;
+    }
 #ifdef DOLBY_UDC
     if (list->codecHasQuirk(
                 index, "needs-flush-before-disable")) {
@@ -2532,6 +2539,14 @@ status_t OMXCodec::allocateOutputBuffersFromNativeWindow() {
     ALOGI("NOTE: Overriding minUndequeuedBufs to %d",minUndequeuedBufs);
 #endif
 
+    // XXX: should hold extra two buffers in surface texture back end.
+    // Currently, intel hw overlay need always hold output buffer until
+    // flip to another buffer. So we should hold more buffers to avoid
+    // buffer overwrite by decoder.
+    if (mQuirks & kRequiresHoldExtraBuffers) {
+        minUndequeuedBufs += 2;
+    }
+
     // XXX: Is this the right logic to use?  It's not clear to me what the OMX
     // buffer counts refer to - how do they account for the renderer holding on
     // to buffers?
